File: 1 Python/Leetcode-MinimumNumberOfOperations.py
# 2 versions
# v2 - left to right and right to left lookup

class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        size = len(boxes)
        ansArr = [0] * size

        # left to right
        lCt = 0
        opL = 0
        for i in range(size):
            ansArr[i]+=opL
            if boxes[i]=='1':
                lCt+=1
            opL+=lCt
        
        #right to left
        rCt = 0
        opR = 0
        for j in range(size-1,-1,-1):
            ansArr[j]+=opR
            if boxes[j]=='1':
                rCt+=1
            opR+=rCt
        
        return ansArr


#v1 O(N^2) BruteForce first principles approach

class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        #actually since in one move only one ball can be moved, this can be approached step by step
        indsOf1 = []

        for i in range(len(boxes)):
            if boxes[i]=="1":
                indsOf1.append(i+1)
        
        ansArr = []
        
        for j in range(len(boxes)):
            moves = 0
            for k in indsOf1:
                moves+=abs(k-(j+1))
            ansArr.append(moves)
        
        return ansArr

File: 2 CPP/Leetcode-MinimumNumberOfOperations.cpp
//2 versions

//v2.2 -  using the addition by boolean
//
class Solution {
public:
    vector<int> minOperations(string boxes) {
        int size = boxes.size();
        vector <int> ansArr (size,0);
        int ctL = 0, opL = 0;
        for(int i=0;i<size;++i){
            ansArr[i]+=opL;
            ctL+= (boxes[i]=='1');
            opL+=ctL;
        }
        int ctR = 0, opR = 0;
        for(int j=size-1;j>=0;--j){
            ansArr[j]+=opR;
            ctR+= (boxes[j]=='1');
            opR+=ctR;
        }
        return ansArr;
    }
};


//v2.1  -  left to right and right to left traversal
//

class Solution {
public:
    vector<int> minOperations(string boxes) {
        int size = boxes.size();
        vector <int> ansArr (size,0);
        int ctL = 0, opL = 0;
        for(int i=0;i<size;++i){
            ansArr[i]+=opL;
            if(boxes[i]=='1'){
                ctL+=1;
            }
            opL+=ctL;
        }
        int ctR = 0, opR = 0;
        for(int j=size-1;j>=0;--j){
            ansArr[j]+=opR;
            if(boxes[j]=='1'){
                ctR+=1;
            }
            opR+=ctR;
        }
        return ansArr;
    }
};


//v1 - comparing the index differences in a brute force manner..
class Solution {
public:
    vector<int> minOperations(string boxes) {
        vector <int> indsOf1;
        for(int i=0;i<boxes.size();++i){
            if(boxes[i]=='1'){
                indsOf1.push_back(i+1);
            }
        }

        vector <int> ansArr;

        for(int j =0;j<boxes.size();++j){
            int moves = 0;
            for(int k: indsOf1){
                moves+=abs(k-(j+1));
            }
            ansArr.push_back(moves);
        }
        return ansArr;
    }
};
